### ISSUE TITLE
Unexpected (buggy?) behaviour from actionCreators.setInitial

### ISSUE THREAD

**Issue by joshacheson** [davidkpiano/react-redux-form#1006]
### The Problem

Calling `setInitial` on my form is transforming certain parts of form state to an unexpected shape.

### Steps to Reproduce

Here is code I have for a sample (node, v8.x) application:

```javascript
const util = require('util');
const redux = require('redux');
const createStore = redux.createStore;
const rrf = require('react-redux-form');

const initialState = {
    somePrimitive: 1,
    someArray: []
};

const modelReducer = rrf.modelReducer('test', initialState);
const formReducer = rrf.formReducer('test', initialState);

const store = createStore(
    redux.combineReducers({
        model: modelReducer,
        form: formReducer
    })
);

store.subscribe(() => {
    console.log(util.inspect(store.getState(), false, null));
});

store.dispatch(rrf.actions.change('test.someArray', [1,2,3]));
store.dispatch(rrf.actions.setInitial('test'));
```

I am concerned with my form state for the `someArray` portion of my form. Prior to dispatching `actions.setInitial` (but after `actions.change`) being called, this portion of my form state is sane (apologize for logging format from node, but I think it's readable enough!):

```
'$form':
         { initialValue: [],
           focus: false,
           pending: false,
           pristine: false,
           submitted: false,
           submitFailed: false,
           retouched: false,
           touched: false,
           valid: true,
           validating: false,
           validated: false,
           validity: {},
           errors: {},
           intents: [ { type: 'validate' } ],
           model: 'test.someArray',
           value: [ 1, 2, 3 ] } } } }
```

after dispatching `actions.setInitial`:

```
'$form':
         { '$form':
            { initialValue: [],
              focus: false,
              pending: false,
              pristine: true,
              submitted: false,
              submitFailed: false,
              retouched: false,
              touched: false,
              valid: true,
              validating: false,
              validated: false,
              validity: {},
              errors: {},
              intents: [ { type: 'reset' } ],
              model: 'test.someArray',
              value: [] } } } } }
```

Notice what has happened, my `$form` property now has its own `$form` property. So all in all, I have,

`test.someArray.$form.$form = {...}`

### Expected Behavior

I expect that my state shape stays broadly consistent, as opposed to being drastically altered by a dispatch to `actions.setInitial`.

### Reproducible Code Example

Code linked above. It runs with Node 8.x. I was having problems with codesandbox (it wasn't giving me the proper exports from `react-redux-form`, so I just did this in node. Note that I am using `react-redux-form@1.16.0`.

### Followup Investigation

After experiencing this behaviour, I did some investigating, so I'll stream of thought here with the steps I encountered. It's hard for me to muse about the best possible fix to this issue.

1) A call to `setInitial` will take me here: https://github.com/davidkpiano/react-redux-form/blob/master/src/reducers/form-actions-reducer.js#L331

2) Which, in the case of my form's model, will quickly lead here: https://github.com/davidkpiano/react-redux-form/blob/master/src/reducers/form-actions-reducer.js#L50-L53. This will be run on `formState.someArray` at some point, which at the time contains a `$form` property, passing what appears to be basic duck typing for "is this a sub form". 

3) a `mapValues` occurs. _Each_ value of my `formState` (meaning the `$form` property, and any others) are mapped with `resetFieldState`.

4) `resetFieldState` isn't particularly interesting here, it's just worth mentioning that we end up calling `fieldOrForm`: https://github.com/davidkpiano/react-redux-form/blob/master/src/reducers/form-actions-reducer.js#L33-L37.

5) In `fieldOrForm`, my subform state is mapped again. It's passed to either `createFormState` or `createFieldState`. When `formState.someArray.$form` is passed to `fieldOrForm`, it will pass the check on line 12: https://github.com/davidkpiano/react-redux-form/blob/master/src/utils/create-field.js#L12. Because of this, we will call `createFormState`.

note: I want to highlight before proceeding that at this point, my object which lived at `formState.someArray.$form` is what's being passed to `createFormState` (or, more accurately, it's at least one of the objects being passed to `createFormState`).

6) Inside `createFormState`, we add a `$form` property to my form state object which already lives at a property by that name: https://github.com/davidkpiano/react-redux-form/blob/master/src/utils/create-field.js#L53. So now we have `formState.someArray.$form.$form...`.

### Conclusion

As far as I can tell this is a bug. That said, I accept that I might be mis-using the library, misunderstanding the purpose of `setInitial`, or anything along those lines.

I would be interested in taking out a PR to fix this (if this does turn out to be a bug). With that said, I'm not sure exactly what the best path would be. As far as I can tell we need to know when we're working with an actual `$form` object and NOT do all of these mappings in this case. I sense perhaps that this was the purpose of https://github.com/davidkpiano/react-redux-form/blob/master/src/reducers/form-actions-reducer.js#L43-L48. However, the `key` prop here never seems to be correct. It's _never_ `$form`.  Instead, `key` here always refers to an actual object. If I'm right about this being the issue, my guess is the bug lies in https://github.com/davidkpiano/react-redux-form/blob/master/src/utils/update-field.js#L38-L77.

Please let me know your thoughts and whether or not I can help.

**Comment by davidkpiano** [comment]
Can you do me a favor and put that in the CodePen or CodeSandbox template? I'll investigate, but I think what `setInitial` should do here is completely replace the `test` form tree instead of recursively trying to update the children.

Thanks for the detailed investigation, by the way!

**Comment by joshacheson** [comment]
Hey there! Is there anything I can do to help move this along?

**Comment by davidkpiano** [comment]
If you have a potential fix, please create a PR for it. I'd be happy to look at it and merge it in ASAP.

**Comment by joshacheson** [comment]
First pass at a PR: https://github.com/davidkpiano/react-redux-form/pull/1024

**Comment by joshacheson** [comment]
Closed by #1024
